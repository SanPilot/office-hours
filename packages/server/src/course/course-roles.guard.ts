import { Injectable, CanActivate, ExecutionContext } from '@nestjs/common';
import { Reflector } from '@nestjs/core';
import { UserModel } from '../profile/user.entity';

@Injectable()
export class RolesGuard implements CanActivate {
  constructor(private reflector: Reflector) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const roles = this.reflector.get<string[]>('roles', context.getHandler());
    if (!roles) {
      return true;
    }
    const request = context.switchToHttp().getRequest();
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    const courseId = request.params.id;
    return this.matchRoles(roles, user, courseId);
  }

  matchRoles(roles: string[], user: UserModel, courseId: number): boolean {
    const userCourse = user.courses.find((course) => {
      return Number(course.courseId) === Number(courseId);
    });

    if (!userCourse) {
      return false;
    }

    const remaining = roles.filter((role) => {
      return userCourse.role.toString() === role;
    });

    return remaining.length > 0;
  }
}
