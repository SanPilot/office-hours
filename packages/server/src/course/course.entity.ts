import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  OneToMany,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { OfficeHourModel } from './office-hour.entity';
import { QueueModel } from '../queue/queue.entity';
import { UserCourseModel } from '../profile/user-course.entity';
import { SemesterModel } from './semester.entity';
import { Exclude } from 'class-transformer';
import { Heatmap } from '@koh/common';

/**
 * Represents a course in the context of office hours.
 * @param id - The id number of this Course.
 * @param name - The subject and course number of this course. Ex: "CS 2500"
 * @param semester - The semester of this course.
 */
/*interface Course {
    id: number;
    name: string;
    url: string;
    semester: Semester;
    users: UserCourse[]
}*/

@Entity('course_model')
export class CourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToMany((type) => OfficeHourModel, (oh) => oh.course)
  officeHours: OfficeHourModel[];

  @OneToMany((type) => QueueModel, (q) => q.course)
  queues: QueueModel[];

  @Column('text')
  name: string;

  @Column('text', { nullable: true })
  @Exclude()
  icalURL: string;

  @OneToMany((type) => UserCourseModel, (ucm) => ucm.course)
  @Exclude()
  userCourses: UserCourseModel;

  @ManyToOne((type) => SemesterModel, (semester) => semester.courses)
  @JoinColumn({ name: 'semesterId' })
  @Exclude()
  semester: SemesterModel;

  @Column({ nullable: true })
  @Exclude()
  // TODO: can we make these not nullable and work with TypeORM
  semesterId: number;

  @Column('boolean', { nullable: true })
  enabled: boolean; // Set to true if the given the course is using our app

  // The heatmap is false when there havent been any questions asked yet or there havent been any office hours
  heatmap: Heatmap | false;
}
