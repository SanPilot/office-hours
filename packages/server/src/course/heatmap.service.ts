import { ClosedQuestionStatus, Heatmap, timeDiffInMins } from '@koh/common';
import { Injectable } from '@nestjs/common';
import { mean } from 'lodash';
import moment = require('moment');
import { Command } from 'nestjs-command';
import { QuestionModel } from 'question/question.entity';

@Injectable()
export class HeatmapService {
  async getHeatmapFor(courseId: number): Promise<Heatmap> {
    // The number of minutes to average across
    const BUCKET_SIZE_IN_MINS = 60;
    // Number of samples to gather per bucket
    const SAMPLES_PER_BUCKET = 5;
    console.time('heatmap');
    const questions = await QuestionModel.createQueryBuilder('question')
      .leftJoinAndSelect('question.queue', 'queue')
      .where('queue.courseId = :courseId', { courseId })
      .andWhere('question.status = :status', {
        status: ClosedQuestionStatus.Resolved,
      })
      .andWhere('question.helpedAt IS NOT NULL')
      .andWhere('question.createdAt > :recent', {
        recent: moment()
          .subtract(8, 'weeks')
          .toISOString(),
      })
      .orderBy('question.createdAt', 'ASC')
      .getMany();

    const heatmap = this._generateHeatMapWithReplay(
      // Ignore questions that cross midnight (usually a fluke)
      questions.filter(q => q.helpedAt.getDate() === q.createdAt.getDate()),
      BUCKET_SIZE_IN_MINS,
      SAMPLES_PER_BUCKET,
    );
    console.timeEnd('heatmap');
    return heatmap;
  }

  // PRIVATE function that is public for testing purposes
  // Rewind through the last few weeks and for each time interval,
  // figure out how long wait time would have been if you had joined the queue at that time
  _generateHeatMapWithReplay(
    questions: QuestionModel[],
    bucketSize: number,
    samplesPerBucket: number,
  ): Heatmap {
    const sampleInterval = bucketSize / samplesPerBucket;
    /*
    TEST: Question1 is  3:05 - 3:25
    // The next question is 3:21 - 3:49
    THe following question is 4:05 - 4:10
    
    Bucket = 60, Samples = 3, so timepoints are: 3:00, 3:20, 3:40.

    3:20 sample gets waittime of 5 minutes
    3:40 samples get waittimes of 9 minutes
    4:00 sample gets waittime of 0 minutes


    If i entered the queue at that time when should I have gotten help?
    Every interval of minutes for the past 5 weeks are aggregated (by taking the avg)
    
    analyze the buckets to find the closest time approximation

    look at question Q1 and the next question Q2
    for all sample timepoints between Q1.createdAt and Q2.createdAt:
       - sample = Q1.closedAt - timepoint (if negative, then it's 0)
    */

    const startDate = questions[0].createdAt;
    const sunday = new Date(startDate);
    sunday.setUTCDate(startDate.getUTCDate() - startDate.getUTCDay());
    sunday.setUTCHours(0, 0, 0, 0);

    function getNextTimepointIndex(date: Date): number {
      return Math.floor(timeDiffInMins(date, sunday) / sampleInterval) + 1;
    }

    // Get the date of the sample timepoint immediately after the given date
    function getNextSampleTimepoint(date: Date): Date {
      const timepointIndex = getNextTimepointIndex(date);
      return new Date(
        sunday.getTime() + timepointIndex * sampleInterval * 60 * 1000,
      );
    }

    function getSampleTimepointsInDateRange(date1: Date, date2: Date): Date[] {
      const ret = [];
      let curr = getNextSampleTimepoint(date1);
      while (curr.getTime() < date2.getTime()) {
        ret.push(curr);
        curr = getNextSampleTimepoint(curr);
      }
      return ret;
    }

    const allTimepointWaitTimes: number[] = [];

    // go two questions at a time
    for (let i = 0; i < questions.length - 1; i++) {
      const curr = questions[i];
      const next = questions[i + 1];

      // get the timepoints in between
      const sampledTimepoints = getSampleTimepointsInDateRange(
        curr.createdAt,
        next.createdAt,
      );

      // When we would have gotten help at this timepoint
      for (const c of sampledTimepoints) {
        allTimepointWaitTimes[getNextTimepointIndex(c)] = Math.max(
          0,
          (curr.helpedAt.getTime() - c.getTime()) / 60000,
        );
      }
    }

    // Bucket the timepoints
    const SAMPLES_PER_HOUR = 60 / sampleInterval;
    const SAMPLES_PER_DAY = 24 * SAMPLES_PER_HOUR;
    const SAMPLES_PER_WEEK = 7 * SAMPLES_PER_DAY;
    const heatmap = [...Array(7 * 24)];
    // for 24 hrs 7 days
    for (let day = 0; day < 7; day++) {
      for (let hour = 0; hour < 24; hour++) {
        const times = [];
        // Iterate through each week
        let week = 0;
        // index in alltimepoints array
        let index =
          week * SAMPLES_PER_WEEK +
          day * SAMPLES_PER_DAY +
          SAMPLES_PER_HOUR * hour;
        while (index < allTimepointWaitTimes.length) {
          for (
            let s = 0;
            s < sampleInterval && index + s < allTimepointWaitTimes.length;
            s++
          ) {
            // Grab all the samples in this bucket from current week
            const t = allTimepointWaitTimes[index + s];
            if (t) {
              times.push(t);
            }
          }
          week++;
          index =
            week * SAMPLES_PER_WEEK +
            day * SAMPLES_PER_DAY +
            SAMPLES_PER_HOUR * hour;
        }
        // times.sort();
        // const median = times[Math.floor(times.length / 2)];
        heatmap[day * 7 + hour] = mean(times);
      }
    }

    return heatmap;
  }

  @Command({
    command: 'heatmap:generate',
    describe: 'generate heatmap for a course',
    autoExit: true,
  })
  async create(): Promise<void> {
    const h = await this.getHeatmapFor(4);
  }
}
