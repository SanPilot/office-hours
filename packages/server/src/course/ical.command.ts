import { Command, Positional } from 'nestjs-command';
import { Injectable } from '@nestjs/common';
import { IcalService } from './ical.service';

@Injectable()
export class ICalCommand {
  constructor(private readonly icalService: IcalService) {}
  @Command({
    command: 'ical:scrape',
    describe: 'scrape ical for a course',
    autoExit: true,
  })
  async create(): Promise<void> {
    await this.icalService.updateAllCourses();
  }
}
