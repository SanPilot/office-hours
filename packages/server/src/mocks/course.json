{
  "schedule": [
    {
      "id": 84,
      "course_name": "CS 2500",
      "course_id": 169,
      "room": "RY 154",
      "ta_name": "Will Stenzel",
      "ta_photo": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=24f4b12cccbf875c7740bbfed45a993900cf0d08d11aa07c84780b3a3501f3bacca4eb33ed5effee8aa2dd195750cfbc9884dd5f2ac62c8f",
      "campus": 116,
      "semester": "Fall 2019",
      "queue": null,
      "start": "2019-09-21T12:00:00-04:00",
      "end": "2019-09-21T14:00:00-04:00",
      "swap": null,
      "canceled": false,
      "ta_arrived_on": null,
      "ta_departed_on": null
    },
    {
      "id": 82,
      "course_name": "CS 2500",
      "course_id": 169,
      "room": "KA 110",
      "ta_name": "Grigory Zaytsev",
      "ta_photo": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=471f2d695fbb8a00ee740ad3ea910453986aec81ddaecf889ae98b3a1858597b12650afd0d4e59c561172f76cb1946eec217ed89bd4074c0",
      "campus": 116,
      "semester": "Fall 2019",
      "queue": null,
      "start": "2019-09-04T06:00:00-04:00",
      "end": "2019-09-04T08:00:00-04:00",
      "swap": null,
      "canceled": false,
      "ta_arrived_on": null,
      "ta_departed_on": null
    }
  ],
  "queues": ["45"]
}
