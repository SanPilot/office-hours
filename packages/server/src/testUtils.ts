import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
