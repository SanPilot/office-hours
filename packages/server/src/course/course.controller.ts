import {
  ClassSerializerInterceptor,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  UseGuards,
  UseInterceptors,
  InternalServerErrorException,
} from '@nestjs/common';
import {
  GetCourseResponse,
  QueuePartial,
  Role,
  TACheckoutResponse,
} from '@koh/common';
import async from 'async';
import { Connection, getRepository, MoreThanOrEqual } from 'typeorm';
import { EventModel, EventType } from 'profile/event-model.entity';
import { JwtAuthGuard } from '../login/jwt-auth.guard';
import { Roles } from '../profile/roles.decorator';
import { User } from '../profile/user.decorator';
import { UserModel } from '../profile/user.entity';
import { QueueCleanService } from '../queue/queue-clean/queue-clean.service';
import { QueueModel } from '../queue/queue.entity';
import { CourseRolesGuard } from './course-roles.guard';
import { CourseModel } from './course.entity';
import { HeatmapService } from './heatmap.service';
import { OfficeHourModel } from './office-hour.entity';
import { QueueSSEService } from '../queue/queue-sse.service';
import moment = require('moment');

@Controller('courses')
@UseGuards(JwtAuthGuard, CourseRolesGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class CourseController {
  constructor(
    private connection: Connection,
    private queueCleanService: QueueCleanService,
    private queueSSEService: QueueSSEService,
    private heatmapService: HeatmapService,
  ) {}

  @Get(':id')
  @Roles(Role.PROFESSOR, Role.STUDENT, Role.TA)
  async get(@Param('id') id: number): Promise<GetCourseResponse> {
    // TODO: for all course endpoint, check if they're a student or a TA
    const course = await CourseModel.findOne(id, {
      relations: ['queues', 'queues.staffList'],
    });

    // Use raw query for performance (avoid entity instantiation and serialization)
    course.officeHours = await getRepository(OfficeHourModel)
      .createQueryBuilder('oh')
      .select(['id', 'title', `"startTime"`, `"endTime"`])
      .where('oh.courseId = :courseId', { courseId: course.id })
      .getRawMany();
    course.heatmap = await this.heatmapService.getHeatmapFor(id);

    course.queues = await async.filter(
      course.queues,
      async (q) => await q.checkIsOpen(),
    );
    await async.each(course.queues, async (q) => {
      await q.addQueueTimes();
      await q.addQueueSize();
    });

    return course;
  }

  @Post(':id/ta_location/:room')
  @Roles(Role.PROFESSOR, Role.TA)
  async checkIn(
    @Param('id') courseId: number,
    @Param('room') room: string,
    @User() user: UserModel,
  ): Promise<QueuePartial> {
    let queue = await QueueModel.findOne(
      {
        room,
        courseId,
      },
      { relations: ['staffList'] },
    );

    if (!queue) {
      queue = await QueueModel.create({
        room,
        courseId,
        staffList: [],
        questions: [],
        allowQuestions: true,
      }).save();
    }

    if (queue.staffList.length === 0) {
      queue.allowQuestions = true;
    }

    queue.staffList.push(user);
    await queue.save();

    await EventModel.create({
      time: new Date(),
      eventType: EventType.TA_CHECKED_IN,
      user,
      courseId,
    }).save();

    await this.queueSSEService.updateQueue(queue.id);
    return queue;
  }

  @Delete(':id/ta_location/:room')
  @Roles(Role.PROFESSOR, Role.TA)
  async checkOut(
    @Param('id') courseId: number,
    @Param('room') room: string,
    @User() user: UserModel,
  ): Promise<TACheckoutResponse> {
    const queue = await QueueModel.findOne(
      {
        room,
        courseId,
      },
      { relations: ['staffList'] },
    );
    queue.staffList = queue.staffList.filter((e) => e.id !== user.id);
    if (queue.staffList.length === 0) {
      queue.allowQuestions = false;
    }
    await queue.save();

    await EventModel.create({
      time: new Date(),
      eventType: EventType.TA_CHECKED_OUT,
      user,
      courseId,
    }).save();

    const canClearQueue = await this.queueCleanService.shouldCleanQueue(queue);
    let nextOfficeHourTime = null;

    // find out how long until next office hour
    if (canClearQueue) {
      const soon = moment().add(15, 'minutes').toDate();
      const nextOfficeHour = await OfficeHourModel.findOne({
        where: { startTime: MoreThanOrEqual(soon) },
        order: {
          startTime: 'ASC',
        },
      });
      nextOfficeHourTime = nextOfficeHour?.startTime;
    }
    await this.queueSSEService.updateQueue(queue.id);
    return { queueId: queue.id, canClearQueue, nextOfficeHourTime };
  }
}
