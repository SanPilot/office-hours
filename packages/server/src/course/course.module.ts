import { Module } from '@nestjs/common';
import { CourseController } from './course.controller';
import { QueueModule } from '../queue/queue.module';
import { ICalCommand } from './ical.command';
import { IcalService } from './ical.service';

@Module({
  controllers: [CourseController],
  imports: [QueueModule],
  providers: [ICalCommand, IcalService],
})
export class CourseModule {}
