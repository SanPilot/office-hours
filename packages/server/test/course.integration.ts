import { CourseModule } from '../src/course/course.module';
import {
  CourseFactory,
  OfficeHourFactory,
  QueueFactory,
  StudentCourseFactory,
  TACourseFactory,
  UserFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';

describe('Course Integration', () => {
  const supertest = setupIntegrationTest(CourseModule);
  describe('GET /courses/:id', () => {
    it('gets office hours and queues', async () => {
      const course = await CourseFactory.create({
        officeHours: [await OfficeHourFactory.create()],
      });
      const response = await supertest({ userId: 1 })
        .get(`/courses/${course.id}`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  //TODO: make a DSL for testing auth points using Hack your own Language
  describe('POST /courses/:id/ta_location/:room', () => {
    it('checks a TA into an existing queue', async () => {
      const queue = await QueueFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({
        course: queue.course,
        user: ta,
      });

      const response = await supertest({ userId: ta.id })
        .post(`/courses/${queue.course.id}/ta_location/${queue.room}`)
        .expect(201);

      expect(response.body).toMatchSnapshot();
    });

    it("Doesn't allow student to check in", async () => {
      const queue = await QueueFactory.create();
      const student = await UserFactory.create();
      await StudentCourseFactory.create({
        course: queue.course,
        user: student,
      });

      await supertest({ userId: student.id })
        .post(`/courses/${queue.course.id}/ta_location/${queue.room}`)
        .expect(500);
    });

    it('checks TA into a new queue', async () => {
      const ta = await UserFactory.create();
      const tcf = await TACourseFactory.create({
        course: await CourseFactory.create(),
        user: ta,
      });
      const response = await supertest({ userId: ta.id })
        .post(`/courses/${tcf.courseId}/ta_location/The Alamo`)
        .expect(201);

      expect(response.body).toMatchObject({
        id: 1,
        room: 'The Alamo',
        staffList: [{ id: ta.id }],
      });
    });
  });
});
