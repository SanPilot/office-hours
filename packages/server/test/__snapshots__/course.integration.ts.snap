// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Integraiton GET /courses/:id gets office hours and queues 1`] = `
Object {
  "id": 1,
  "name": "CS 2500",
  "officeHours": Array [
    Object {
      "endTime": "2020-05-20T15:30:00.000Z",
      "id": 1,
      "room": "WVH 101",
      "startTime": "2020-05-20T14:00:00.000Z",
      "title": "Alex & Stanley",
    },
  ],
  "queues": Array [],
}
`;
